pipeline/execout/interface.go

```diff
@@ -0,0 +1,32 @@
+package execout
+
+import (
+	"context"
+	"errors"
+	"github.com/streamingfast/bstream"
+	pbsubstreams "github.com/streamingfast/substreams/pb/sf/substreams/v1"
+)
+
+var NotFound = errors.New("inputs module value not found")
+
+type CacheEngine interface {
+	NewExecOutput(blockType string, block *bstream.Block, clock *pbsubstreams.Clock, cursor *bstream.Cursor) (ExecOutput, error)
+	Prime(modules map[string]*pbsubstreams.Module) error
+	UndoBlock(moduleName string, blockID string)
+	Flush(ctx context.Context) error
+	FlushAndUpdate(ctx context.Context, blockRef bstream.BlockRef) error
```

Comment:
the caller ought not to know what the CacheEngine will do with this information, if it wants to Flush or not.. it's not the caller's concern.

If we have `UndoBlock` and that's the minimum thing we want to instruct the cache engine, then we should have something symmetric for the other block steps, like `NewBlock()` .. and let the engine decide whether to flush or not. It seems that that's what `FlushAndUpdate()` wants to be, doesn't it?

Reply:
I'm unsure of the passing of `ctx context.Context` here also. The CacheEngine deals with its own context.. for example, you wouldn't want to cancel a flush because someone suddenly disconnects, right? The flush is not related to a particular request.

Reply:
as for `Flush()` .. again it's not the concern of the caller what the action should be. The message being passed looks more like an `RangeEndReached()` .. which would come with a number that potentially is over a boundary, but has not been processed.. so it could inform the caching engine to flush what happened before.

So I imagine:

```
NewExecOutput()
Init(ModuleHashes)
NewBlock(blockRef)
UndoBlock(blockID string) // let UndoBlock do the looping
RangeEndReached(blockNum uint64)
```

Reply:
There's also the issue that `RangeEndReached()` does not guarantee things to be irreversible, so shouldn't _necessarily_ trigger a Flush(), unless perhaps we also pass down the `Step`, and make a decision based on that.  See comments near `flushCache()`.